README.md
<h1 align="center">
  <span style="font-size:24px">🌿🕸️ Mongo Data API</span><br />
  <sub><i>Mongo Client for an HTTPS World. Lambda and edge function friendly</i></sub>
</h1>

<!-- Shields -->
<a href="https://www.npmjs.com/package/@taskless/mongo-data-api"><img alt="npm" src="https://img.shields.io/npm/v/%40taskless/mongo-data-api"></a>&nbsp;
<a href="https://www.npmjs.com/package/@taskless/mongo-data-api"><img alt="npm" src="https://img.shields.io/npm/dm/%40taskless/mongo-data-api"></a>&nbsp;
<a href="https://github.com/taskless/mongo-data-api"><img alt="GitHub issues" src="https://img.shields.io/github/issues/taskless/mongo-data-api"></a>&nbsp;
<a href="https://github.com/taskless/mongo-data-api"><img alt="GitHub" src="https://img.shields.io/github/license/taskless/mongo-data-api"></a>&nbsp;


> A Mongo-like API for accessing the http-based [Mongo Data API](https://www.mongodb.com/docs/atlas/api/data-api/). Uses [BSON](https://www.npmjs.com/package/bson) to provide access to standard Mongo data types. BYO `fetch()` for maximum portability.

- [Usage](#usage)
  - [Authentication](#authentication)
    - [Using a Data API api key (preferred)](#using-a-data-api-api-key-preferred)
    - [Using Email and Password](#using-email-and-password)
    - [Using a custom JWT](#using-a-custom-jwt)
    - [Using Bearer Auth (required for browser)](#using-bearer-auth-required-for-browser)
- [Supported Methods and API](#supported-methods-and-api)
  - [Create a Mongo Client](#create-a-mongo-client)
  - [Select a Database](#select-a-database)
  - [Select a Collection](#select-a-collection)
  - [Changing `fetch()` on the Fly](#changing-fetch-on-the-fly)
  - [Collection Methods](#collection-methods)
    - [Return Type](#return-type)
    - [Methods](#methods)
      - [findOne](#findone)
      - [find](#find)
      - [insertOne](#insertone)
      - [insertMany](#insertmany)
      - [updateOne](#updateone)
      - [updateMany](#updatemany)
      - [replaceOne](#replaceone)
      - [deleteOne](#deleteone)
      - [deleteMany](#deletemany)
      - [aggregate](#aggregate)
      - [callApi](#callapi)
- [Errors](#errors)
- [FAQ](#faq)
- [License](#license)

# Usage

```ts
import { MongoClient } from "@taskless/mongo-data-api";

const mc = new MongoClient({
  /** @type URL | string */
  endpoint: new URL(process.env.MONGO_HTTP_URL ?? "[not defined]"),
  /** @type string */
  dataSource: process.env.MONGO_HTTP_DATA_SOURCE ?? "[not defined]",

  /* See "Authentication" below */
  auth: {
    /* ... */
  },
});

const { data, error } = await mc.findOne({
  /* good 'ole mongo! See the Collection Methods for what's available */
});
```

## Authentication

| Authentication Method | Supported |
| :-------------------- | :-------: |
| API Key               |    ✅     |
| Email & Password      |    ✅     |
| Custom JWT            |    ✅     |
| Bearer                |    ⚠️     |

### Using a Data API api key (preferred)

```ts
{
  // ...
  auth: {
    /** @type string */
    apiKey: process.env.MONGO_HTTP_API_KEY ?? "[not defined]",
  },
}
```

### Using Email and Password

```ts
{
  // ...
  auth: {
    /** @type string */
    email: process.env.MONGO_EMAIL ?? "",
    /** @type string */
    password: process.env.MONGO_PASSWORD ?? "",
  },
}
```

### Using a custom JWT

```ts
{
  // ...
  auth: {
    /** @type string */
    jwtTokenString: request.headers.get("jwt"),
  },
}
```

### Using Bearer Auth (required for browser)

_[Read more about authenticating Realm users in the browser](https://www.mongodb.com/docs/atlas/app-services/users/sessions/#std-label-manage-user-sessions)_

```ts
{
  // ...
  auth: {
    /** @type string */
    bearerToken: tokenFromRealm,
  },
}
```

# Supported Methods and API

## Create a Mongo Client

```ts
const client = new MongoClient(options);
```

- `options` - MongoClient options
  - `options.endpoint` - `string | URL` an endpoint for sending requests to. Your Data API Endpoint is available at your Mongo Data API UI `https://cloud.mongodb.com/v2/<projectId>#/dataAPI`, where `<projectId>` is your project ID. A single Data API is usable for the entire project, with individual data sources routing to specific atlas instances.
  - `options.dataSource` - `string` the `Data Source` for your Data API. On the Data API UI, this is the "Data Source" column, and usually is either a 1:1 mapping of your cluster name, or the default `mongodb-atlas` if you enabled Data API through the Atlas Admin UI.
  - `options.auth` - `AuthOptions` one of the authentication methods, either api key, email & password, or a custom JWT string. At this time, only [Credential Authentication](https://www.mongodb.com/docs/atlas/api/data-api/#credential-authentication) is supported.
  - `options.fetch?` - A custom `fetch` function conforming to the [native fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API). We recommend [cross-fetch](https://www.npmjs.com/package/cross-fetch), as it asserts a complaint `fetch()` interface and avoids you having to do `fetch: _fetch as typeof fetch` to satisfy the TypeScript compiler
  - `options.headers?` - Additional [Headers](https://developer.mozilla.org/en-US/docs/Web/API/Headers) to include with the request. May also be a simple object of key/value pairs.

## Select a Database

```ts
const db = client.db(databaseName);
```

- `databaseName` - `string` the name of the database to connect to

## Select a Collection

```ts
const collection = db.collection<TSchema>(collectionName);
```

- `collectionName` - `string` the name of the collection to connect to
- `<TSchema>` - _generic_ A Type or Interface that describes the documents in this collection. Defaults to the generic MongoDB `Document` type

## Changing `fetch()` on the Fly

```ts
const altDb = db.fetch(altFetch);
const altCollection = db.collection<TSchema>(collectionName).fetch(altFetch);
```

- `altFetch` - An alternate `fetch` implementation that will be used for that point forward. Useful for adding or removing retry support on a per-call level, changing the authentication required, or other fetch middleware operations.

## Collection Methods

The following [Data API resources](https://www.mongodb.com/docs/atlas/api/data-api-resources/) are supported

| resource     | support |
| :----------- | :-----: |
| `findOne`    |   ✅    |
| `find`       |   ✅    |
| `insertOne`  |   ✅    |
| `insertMany` |   ✅    |
| `updateOne`  |   ✅    |
| `updateMany` |   ✅    |
| `replaceOne` |   ✅    |
| `deleteOne`  |   ✅    |
| `deleteMany` |   ✅    |
| `aggregate`  |   ✅    |

Should Data API add support for other resources, the `callApi` method allows you to pass arbitrary JSON to a Data API resource. `mongo-data-api` automatically merges the `dataSource`, `database`, and `collection` parameters in if not specified.

### Return Type

All collection methods return an object containing `data?` and `error?`. This avoids throwing during requests, making it easier to handle the response without nesting try/catch operations.

```ts
interface DataAPIResponse {
  data?: TSchema;
  error?: DataAPIError;
}
```

### Methods

#### findOne

```ts
const { data, error } = await collection.findOne(filter, options);
```

- `filter?` - `Filter<TSchema>` A [MongoDB Query Filter](https://www.mongodb.com/docs/manual/tutorial/query-documents/)
- `options?` - Query options
  - `options.projection?` - `Document` A [MongoDB Query Projection](https://www.mongodb.com/docs/manual/tutorial/project-fields-from-query-results/)

#### find

```ts
const { data, error } = await collection.find(filter, options);
```

- `filter?` - `Filter<TSchema>` A [MongoDB Query Filter](https://www.mongodb.com/docs/manual/tutorial/query-documents/)
- `options?` Query options
  - `options.projection?` - `Document` A [MongoDB Query Projection](https://www.mongodb.com/docs/manual/tutorial/project-fields-from-query-results/)
  - `options.sort?` - `Sort` A [MongoDB Sort Expression](https://www.mongodb.com/docs/manual/reference/operator/aggregation/sort/)
  - `options.limit?` - `number` The maximum number of matched documents to include in the returned result set. Each request may return up to 50,000 documents.
  - `options.skip?` - `number` The number of matched documents to skip before adding matched documents to the result set.

#### insertOne

```ts
const { data, error } = await collection.insertOne(document);
```

- `document` - `TSchema` The document to insert

#### insertMany

```ts
const { data, error } = await collection.insertMany(documents);
```

- `documents` - `TSchema[]` The documents to insert

#### updateOne

```ts
const { data, error } = await collection.updateOne(filter, update, options);
```

- `filter` - `Filter<TSchema>` A [MongoDB Query Filter](https://www.mongodb.com/docs/manual/tutorial/query-documents/)
- `update` - `UpdateFilter<TSchema> | Partial<TSchema>` A [MongoDB Update Expression](https://www.mongodb.com/docs/manual/tutorial/update-documents/) that specifies how to modify the matched document
- `options?` Query options
  - `options.upsert` - `boolean` The upsert flag only applies if no documents match the specified filter. If true, the updateOne action inserts a new document that matches the filter with the specified update applied to it.

#### updateMany

```ts
const { data, error } = await collection.updateMany(filter, update, options);
```

- `filter` - `Filter<TSchema>` A [MongoDB Query Filter](https://www.mongodb.com/docs/manual/tutorial/query-documents/)
- `update` - `UpdateFilter<TSchema> | Partial<TSchema>` A [MongoDB Update Expression](https://www.mongodb.com/docs/manual/tutorial/update-documents/) that specifies how to modify the matched document
- `options?` Query options
  - `options.upsert` - `boolean` The upsert flag only applies if no documents match the specified filter. If true, the updateOne action inserts a new document that matches the filter with the specified update applied to it.

#### replaceOne

```ts
const { data, error } = await collection.replaceOne(
  filter,
  replacement,
  options
);
```

- `filter` - `Filter<TSchema>` A [MongoDB Query Filter](https://www.mongodb.com/docs/manual/tutorial/query-documents/)
- `replacement` - `WithoutId<TSchema>` The replacement document, without an `_id` attribute
- `options?` Query options
  - `options.upsert` - `boolean` The upsert flag only applies if no documents match the specified filter. If true, the updateOne action inserts a new document that matches the filter with the specified update applied to it.

#### deleteOne

```ts
const { data, error } = await collection.deleteOne(filter);
```

- `filter` - `Filter<TSchema>` A [MongoDB Query Filter](https://www.mongodb.com/docs/manual/tutorial/query-documents/)

#### deleteMany

```ts
const { data, error } = await collection.deleteMany(filter);
```

- `filter` - `Filter<TSchema>` A [MongoDB Query Filter](https://www.mongodb.com/docs/manual/tutorial/query-documents/)

#### aggregate

```ts
const { data, error } = await collection.aggregate<TOutput>(pipeline);
```

- `pipeline` - `Document[]` A [MongoDB Aggregation Pipeline](https://www.mongodb.com/docs/manual/core/aggregation-pipeline/)
- `<TOutput>` - _generic_ A `Document` like object that describes the output of the aggregation pipeline

#### callApi

```ts
const { data, error } = await collection.callApi<T>(method, body);
```

- `method` - `string` A supported Mongo Data API Request method
- `body` - `Record<string, unknown>` An arbitrary key/value JSON-like data structure representing the body payload sent to the Mongo Data API
- `<T>` - _generic_ Describes the return type of `data` on a successful API call

# Errors

Requests via `fetch()` have their resposne codes checked against the [Data API Error Codes](https://www.mongodb.com/docs/atlas/api/data-api-resources/#error-codes) and on error, set the `error` property of the response to a `MongoDataAPIError`.

- `error.code` - `number` Contains the HTTP error code from the Mongo Data API

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Breaking Changes

### Added

### Fixed

### Changed

### Removed

## [0.4.0]

### Added

- Custom fetch: For `db()` and `collection()`, there is now a chainable `fetch()` method, which returns a new DB or collection with its `fetch` implementation changed. This allows you to implement your own custom logic for retries, without the addition of pRetry

### Removed

- Removed pRetry dependency in favor of users overriding via their own fetch function

## [0.3.0]

### Breaking Changes

- Updated minimum node version to `>=18.0.0`
- Package is now ESM-only. One major reason for this change is that TypeScript now understands ESM importing naturally, making it easy to use this package in TypeScript projects without any additional configuration. If you are using CommonJS, you will need to use a bundler such as Webpack or Rollup to use this package, or import the library dynamically using `await import("mongo-data-api")`. The following snippet will provide support for importing your module safely in CommonJS.

```ts
let mongoDataApiClient: MongoClient | undefined;
const createMongoClient = async () => {
  if (mongoDataApiClient) return mongoDataApiClient;

  const { MongoClient } = await import("mongo-data-api");
  mongoDataApiClient = new MongoClient(/*options*/);
  return mongoDataApiClient;
};
```

### Added

- Smart retries via `p-retry` for hard failures such as `ECONNRESET` and `ECONNREFUSED`

### Fixed

- Return type on `insertOne` is now of type `ObjectId` instead of `string`

## [0.2.2]

### Fixed

- `undefined` values in top level fields (sort, projection, etc) cause `400` from mongo #1

## [0.2.1] - 2023-05-02

### Fixed

- Updated typescript documentation for `fetch?`

## [0.2.0] - 2023-05-02

### Breaking Changes

##### `fetch` is now a standard fetch function, and no longer an object requiring `Request`, `Response`, and `Headers`

When providing a custom `fetch()` function, it's no longer required to also pass the other elements such as `Request`, `Response` and `Headers`. This increases the number of libraries you can use with fetch.

```ts
// OLD
const c = new MongoClient({
  endpoint: BASE_URL,
  dataSource: "test-datasource",
  auth: { apiKey: "validApiKey" },
  fetch: {
    // fetch interface
  },
});

// NEW
const c = new MongoClient({
  endpoint: BASE_URL,
  dataSource: "test-datasource",
  auth: { apiKey: "validApiKey" },
  fetch: ponyfillFetch, // "as typeof fetch" may be required for some fetch libraries
});
```

### Added

- Created a `CHANGELOG.md` for changelog tracking
- Enabled support for plain JS objects in `headers`

### Fixed

### Changed

- Removed requirement for `Request` and `Response` in the custom fetch interface

### Removed

- Dropped `node-fetch` for `cross-fetch` in tests to minimze type assertions

## [0.1.2] - 2023-05-01

### Added

- Mongo Data API (all methods)
- Mongo Data API Auth
- Documentation
- Tests
- Standardized Release Process

## Older Releases

Older releases are available via github releases: https://github.com/taskless/mongo-data-api/releases

<!-- Releases -->

[unreleased]: https://github.com/taskless/mongo-data-api/compare/0.4.0...HEAD
[0.4.0]: https://github.com/taskless/mongo-data-api/compare/0.3.0...0.4.0
[0.3.0]: https://github.com/taskless/mongo-data-api/compare/0.2.2...0.3.0
[0.2.2]: https://github.com/taskless/mongo-data-api/compare/0.2.1...0.2.2
[0.2.1]: https://github.com/taskless/mongo-data-api/compare/0.2.0...0.2.1
[0.2.0]: https://github.com/taskless/mongo-data-api/compare/0.1.2...0.2.0
[0.1.2]: https://github.com/taskless/mongo-data-api/compare/320744af834ca94e450e2a129283e5c7500b763d...0.1.2

<!--
Template:

### Breaking Changes
### Added
### Fixed
### Changed
### Removed
-->


src/client.ts
import { type OutgoingHttpHeaders } from "node:http";
import { EJSON, type ObjectId } from "bson";
import type {
  Sort,
  Filter,
  OptionalUnlessRequiredId,
  UpdateFilter,
  WithId,
  WithoutId,
  Document,
} from "mongodb";
import type { AuthOptions } from "./authTypes.js";
import { MongoDataAPIError } from "./errors.js";

// reexport bson types
export { EJSON, ObjectId } from "bson";

// reexport relevant mongo types
export type {
  Document,
  Sort,
  Filter,
  OptionalUnlessRequiredId,
  UpdateFilter,
  WithId,
  WithoutId,
} from "mongodb";

// eslint-disable-next-line @typescript-eslint/ban-types
type Nullable<T> = T | null;

/** A data API response object */
type DataAPIResponse<T> = { data?: T; error?: MongoDataAPIError };

export type MongoClientConstructorOptions = {
  /**
   * Your atlas endpoint. Usually in the form of
   * https://data.mongodb-api.com/app/<clientAppId>/endpoint/data/v1
   * and retrievable from the Data API UI in Atlas. `clientAppId` is your Atlas App Service
   * friendly ID (not the hex string `_id` value).
   *
   * Your endpoint information is available at https://cloud.mongodb.com/v2/<projectId>#/dataAPI
   */
  endpoint: URL | string;
  /**
   * The [Atlas Data Source](https://cloud.mongodb.com/v2/<projectId>#/dataAPI). This tells
   * mongo which Cluster you want to run your queries against. The default data source
   * usually mimics your cluster's name, but you can find it in the Atlas UI.
   */
  dataSource: string;
  /**
   * An authentication method to use for your Atlas App Service. Supports API Key,
   * Email/Password, and Custom JWT as authentication methods.
   *
   * Read more: https://www.mongodb.com/docs/atlas/api/data-api/#authenticate-requests
   */
  auth: AuthOptions;
  /**
   * Provide a compatibility layer with a custom fetch implementation.
   */
  fetch?: typeof fetch;
  /**
   * Provide a custom Headers object for the request. In some situations such as testing,
   * this allows you to control the response from the mock server.
   */
  headers?: OutgoingHttpHeaders | Headers;
};

type ConnectionOptions = {
  dataSource: string;
  endpoint: string;
  headers: Record<string, string>;
};

type ConnectionOptionsWithDatabase = ConnectionOptions & {
  database: string;
};

type ConnectionOptionsWithCollection = ConnectionOptionsWithDatabase & {
  collection: string;
};

/**
 * Removes empty keys from an object at the top level. EJSON.stringify does not drop
 * undefined values in serialization, so we need to explicitly remove any keys with
 * null/undefined without recursing further into the object.
 */
const removeEmptyKeys = (object: Record<string, unknown>) => {
  return Object.fromEntries(Object.entries(object).filter(([_, v]) => v));
};

/**
 * Create a MongoDB-like client for communicating with the Atlas Data API
 *
 * ref: https://www.mongodb.com/docs/atlas/api/data-api-resources/
 *
 * The Data API client operates over https, which requires an available fetch()
 * implementation, either provided via polyfill, ponyfill, or natively available.
 *
 * To maximize compatibility, the client uses the application/ejson format in all
 * requests, converting mongo-style BSON to JSON and back again.
 */
export class MongoClient {
  protected connection: ConnectionOptions;
  protected ftch: typeof fetch;

  constructor({
    auth,
    dataSource,
    endpoint,
    fetch: customFetch,
    headers: h,
  }: MongoClientConstructorOptions) {
    this.ftch = customFetch ?? globalThis.fetch;
    const headers: HeadersInit = {};

    // accept a node-style or whatwg headers object with .keys() and .get()
    if (typeof h?.keys === "function") {
      for (const key of h.keys()) {
        headers[key] = (
          typeof h.get === "function" ? h.get(key) : headers[key]
        )!;
      }
    }

    this.connection = {
      dataSource,
      endpoint: endpoint instanceof URL ? endpoint.toString() : endpoint,
      headers,
    };

    if (!this.ftch || typeof this.ftch !== "function") {
      throw new Error(
        "No viable fetch() found. Please provide a fetch interface"
      );
    }

    this.connection.headers["Content-Type"] = "application/ejson";
    this.connection.headers.Accept = "application/ejson";

    if ("apiKey" in auth) {
      this.connection.headers.apiKey = auth.apiKey;
      return;
    }

    if ("jwtTokenString" in auth) {
      this.connection.headers.jwtTokenString = auth.jwtTokenString;
      return;
    }

    if ("email" in auth && "password" in auth) {
      this.connection.headers.email = auth.email;
      this.connection.headers.password = auth.password;
      return;
    }

    if ("bearerToken" in auth) {
      this.connection.headers.Authorization = `Bearer ${auth.bearerToken}`;
      return;
    }

    throw new Error("Invalid auth options");
  }

  /** Select a database from within the data source */
  db(name: string) {
    return new Database(name, this.connection, this.ftch);
  }
}

/**
 * Represents a Mongo Database object. Contains information about the selected
 * database, as well as connection information for performing
 * queries.
 */
export class Database {
  protected connection: ConnectionOptionsWithDatabase;
  protected ftch: typeof fetch;

  constructor(
    name: string,
    connection: ConnectionOptions,
    customFetch?: typeof fetch
  ) {
    this.ftch = customFetch ?? globalThis.fetch;
    this.connection = {
      ...connection,
      database: name,
    };
  }

  /**
   * Get a Collection from the Database
   * @param name The collection name
   * @typeParam `T` The type of document stored in the collection
   * @returns A Collection object of type `T`
   */
  collection<TSchema = Document>(name: string) {
    return new Collection<TSchema>(name, this.connection, this.ftch);
  }

  /**
   * Change the fetch interface for this database. Returns a new Database object
   * @param customFetch A fetch interface to use for subsequent calls
   * @returns A new Database object with the custom fetch interface implemented
   */
  fetch(customFetch: typeof fetch) {
    const db = new Database(
      this.connection.database,
      this.connection,
      customFetch
    );
    return db;
  }
}

/**
 * Represents a Mongo Collection object of type `<T>`, providing query
 * methods in a fluent interface.
 */
export class Collection<TSchema = Document> {
  protected connection: ConnectionOptionsWithCollection;
  protected ftch: typeof fetch;

  constructor(
    name: string,
    database: ConnectionOptionsWithDatabase,
    customFetch?: typeof fetch
  ) {
    this.ftch = customFetch ?? globalThis.fetch;
    this.connection = {
      ...database,
      collection: name,
    };
  }

  /**
   * Change the fetch interface for this collection. Returns a new Collection object
   * @param customFetch A fetch interface to use for subsequent calls
   * @returns A new Collection object with the custom fetch interface implemented
   */
  fetch(customFetch: typeof fetch) {
    const collection = new Collection(
      this.connection.collection,
      this.connection,
      customFetch
    );
    return collection;
  }

  /**
   * Find a single document in the collection. **note:** The `findOne()` operation does not provide a sort operation for finding a returning the first record in a set. If you must `findOne` and `sort`, use the `find()` operation instead.
   * [example](https://www.mongodb.com/docs/atlas/api/data-api-resources/#find-a-single-document)
   *
   * @param filter A [MongoDB Query Filter](https://www.mongodb.com/docs/manual/tutorial/query-documents/). The findOne action returns the first document in the collection that matches this filter.
   *
   * If you do not specify a filter, the action matches all document in the collection.
   * @param options Query options
   * @param options.projection A [MongoDB Query Projection](https://www.mongodb.com/docs/manual/tutorial/project-fields-from-query-results/). Depending on the projection, the returned document will either omit specific fields or include only specified fields or values.
   * @returns A matching document, or `null` if no document is found
   */
  async findOne(
    filter?: Filter<TSchema>,
    options: { projection?: Document; sort?: Document } = {}
  ): Promise<DataAPIResponse<Nullable<WithId<TSchema>>>> {
    const { data, error } = await this.callApi<{ document: WithId<TSchema> }>(
      "findOne",
      {
        filter,
        projection: options.projection,
        sort: options.sort,
      }
    );

    if (data) {
      return { data: data.document ?? null, error };
    }

    return { data, error };
  }

  /**
   * Find multiple documents in the collection matching a query.
   * [example](https://www.mongodb.com/docs/atlas/api/data-api-resources/#find-multiple-documents)
   *
   * @param filter A [MongoDB Query Filter](https://www.mongodb.com/docs/manual/tutorial/query-documents/). The find action returns documents in the collection that match this filter.
   *
   * If you do not specify a `filter`, the action matches all document in the collection.
   *
   * If the filter matches more documents than the specified `limit`, the action only returns a subset of them. You can use `skip` in subsequent queries to return later documents in the result set.
   * @param options Query options
   * @param options.projection A [MongoDB Query Projection](https://www.mongodb.com/docs/manual/tutorial/project-fields-from-query-results/). Depending on the projection, the returned document will either omit specific fields or include only specified fields or values.
   * @param options.sort A [MongoDB Sort Expression](https://www.mongodb.com/docs/manual/reference/operator/aggregation/sort/). Matched documents are returned in ascending or descending order of the fields specified in the expression.
   * @param options.limit The maximum number of matched documents to include in the returned result set. Each request may return up to 50,000 documents.
   * @param options.skip The number of matched documents to skip before adding matched documents to the result set.
   * @returns An array of matching documents
   */
  async find(
    filter?: Filter<TSchema>,
    options?: {
      projection?: Document;
      sort?: Sort;
      limit?: number;
      skip?: number;
    }
  ): Promise<DataAPIResponse<WithId<TSchema>[]>> {
    const { data, error } = await this.callApi<{
      documents: WithId<TSchema>[];
    }>("find", {
      filter,
      projection: options?.projection,
      sort: options?.sort,
      limit: options?.limit,
      skip: options?.skip,
    });

    if (data) {
      return { data: data.documents, error };
    }

    return { data, error };
  }

  /**
   * Insert a single document into the collection. [read more](https://www.mongodb.com/docs/atlas/api/data-api-resources/#insert-a-single-document)
   * @param document The document to insert (of type `T`)
   * @returns The insertOne action returns the _id value of the inserted document as a string in the insertedId field
   */
  async insertOne(
    document: OptionalUnlessRequiredId<TSchema>
  ): Promise<DataAPIResponse<{ insertedId: ObjectId }>> {
    return this.callApi("insertOne", {
      document,
    });
  }

  /**
   * Insert several documents into the collection. [read more](https://www.mongodb.com/docs/atlas/api/data-api-resources/#insert-multiple-documents)
   * @param docs An array of documents to insert (of type `T[]`)
   * @returns The insertMany action returns the _id values of all inserted documents as an array of strings in the insertedIds field
   */
  async insertMany(
    documents: OptionalUnlessRequiredId<TSchema>[]
  ): Promise<DataAPIResponse<{ insertedIds: string[] }>> {
    return this.callApi("insertMany", { documents });
  }

  /**
   * Update a single matching document in the collection. [read more](https://www.mongodb.com/docs/atlas/api/data-api-resources/#update-a-single-document)
   * @param filter A [MongoDB Query Filter](https://www.mongodb.com/docs/manual/tutorial/query-documents/). The updateOne action modifies the first document in the collection that matches this filter.
   * @param update A [MongoDB Update Expression](https://www.mongodb.com/docs/manual/tutorial/update-documents/) that specifies how to modify the matched document.
   * @param options Query options
   * @param options.upsert The upsert flag only applies if no documents match the specified filter. If true, the updateOne action inserts a new document that matches the filter with the specified update applied to it.
   * @returns The updateOne action returns:
   *   - the number of documents that the filter matched in the `matchedCount` field
   *   - the number of matching documents that were updated in the `modifiedCount` field
   *
   * If upsert is set to true and no documents match the filter, the action returns the `_id` value of the inserted document as a string in the `upsertedId` field
   */
  async updateOne(
    filter: Filter<TSchema>,
    update: UpdateFilter<TSchema> | Partial<TSchema>,
    options: { upsert?: boolean } = {}
  ): Promise<
    DataAPIResponse<{
      matchedCount: number;
      modifiedCount: number;
      upsertedId?: string;
    }>
  > {
    return this.callApi("updateOne", {
      filter,
      update,
      upsert: options.upsert,
    });
  }

  /**
   * Update multiple documents in the collection matching a criteria. [read more](https://www.mongodb.com/docs/atlas/api/data-api-resources/#update-multiple-documents)
   * @param filter A [MongoDB Query Filter](https://www.mongodb.com/docs/manual/tutorial/query-documents/). The updateMany action modifies all documents in the collection that match this filter.
   * @param update A [MongoDB Update Expression](https://www.mongodb.com/docs/manual/tutorial/update-documents/) that specifies how to modify matched documents.
   * @param options Query options
   * @param options.upsert The upsert flag only applies if no documents match the specified filter. If true, the updateMany action inserts a new document that matches the filter with the specified update applied to it.
   * @returns The updateMany action returns:
   *   - the number of documents that the filter matched in the `matchedCount` field
   *   - the number of matching documents that were updated in the `modifiedCount` field
   *
   * If upsert is set to true and no documents match the filter, the action returns the `_id` value of the inserted document as a string in the `upsertedId` field
   */
  async updateMany(
    filter: Filter<TSchema>,
    update: UpdateFilter<TSchema>,
    { upsert }: { upsert?: boolean } = {}
  ): Promise<
    DataAPIResponse<{
      matchedCount: number;
      modifiedCount: number;
      upsertedId?: string;
    }>
  > {
    return this.callApi("updateMany", {
      filter,
      update,
      upsert,
    });
  }

  /**
   * Replace a single document in the collection. [read more](https://www.mongodb.com/docs/atlas/api/data-api-resources/#replace-a-single-document)
   * @param filter A [MongoDB Query Filter](https://www.mongodb.com/docs/manual/tutorial/query-documents/). The replaceOne action overwrites the first document in the collection that matches this filter.
   * @param replacement The replacement document.
   * @param options Query options
   * @param options.upsert The upsert flag only applies if no documents match the specified filter. If true, the replaceOne action inserts a new document that matches the filter with the specified replacement applied to it.
   * @returns The replaceOne action returns:
   *   - the number of documents that the filter matched in the `matchedCount` field
   *   - the number of matching documents that were updated in the `modifiedCount` field
   *
   * If upsert is set to true and no documents match the filter, the action returns the `_id` value of the inserted document as a string in the `upsertedId` field
   */
  async replaceOne(
    filter: Filter<TSchema>,
    replacement: WithoutId<TSchema>,
    options: { upsert?: boolean } = {}
  ): Promise<
    DataAPIResponse<{
      matchedCount: number;
      modifiedCount: number;
      upsertedId?: string;
    }>
  > {
    return this.callApi("replaceOne", {
      filter,
      replacement,
      upsert: options.upsert,
    });
  }

  /**
   * Delete a single document in the collection. [read more](https://www.mongodb.com/docs/atlas/api/data-api-resources/#delete-a-single-document)
   * @param filter A [MongoDB Query Filter](https://www.mongodb.com/docs/manual/tutorial/query-documents/). The deleteOne action deletes the first document in the collection that matches this filter.
   * @returns The deleteOne action returns the number of deleted documents in the deletedCount field
   */
  async deleteOne(
    filter: Filter<TSchema>
  ): Promise<DataAPIResponse<{ deletedCount: number }>> {
    return this.callApi("deleteOne", {
      filter,
    });
  }

  /**
   * Delete multiple documents from the collection matching a filter. [read more](https://www.mongodb.com/docs/atlas/api/data-api-resources/#delete-multiple-documents)
   * @param filter A [MongoDB Query Filter](https://www.mongodb.com/docs/manual/tutorial/query-documents/). The deleteMany action deletes all documents in the collection that match this filter.
   * @returns The deleteMany action returns the number of deleted documents in the deletedCount field
   */
  async deleteMany(
    filter: Filter<TSchema>
  ): Promise<DataAPIResponse<{ deletedCount: number }>> {
    return this.callApi("deleteMany", { filter });
  }

  /**
   * Run an aggregation pipeline on the collection. [read more](https://www.mongodb.com/docs/atlas/api/data-api-resources/#aggregate-documents)
   * @param pipeline A [MongoDB Aggregation Pipeline](https://www.mongodb.com/docs/manual/core/aggregation-pipeline/).
   * @returns The aggregate action returns the result set of the final stage of the pipeline as an array of documents, automatically unwrapping the Atlas `documents` field
   */
  async aggregate<TOutput = Document>(
    pipeline: Document[]
  ): Promise<DataAPIResponse<TOutput[]>> {
    const response = await this.callApi<{ documents: TOutput[] }>("aggregate", {
      pipeline,
    });

    return response.data
      ? { data: response.data.documents }
      : { error: response.error ?? new MongoDataAPIError("Unknown error", -1) };
  }

  /**
   * Call a raw Data API resource
   * @param method The Data API resource to call
   * @param body The JSON body to include, merged with the collection, dataSource, and database names
   * @returns The response from the Data API, or a DataAPIError if the request failed
   */
  async callApi<T = unknown>(
    method: string,
    body: Record<string, unknown>
  ): Promise<DataAPIResponse<T>> {
    const { endpoint, dataSource, headers, collection, database } =
      this.connection;
    const url = `${endpoint}/action/${method}`;

    const response = await this.ftch(url, {
      method: "POST",
      headers,
      body: EJSON.stringify({
        collection,
        database,
        dataSource,
        ...removeEmptyKeys(body),
      }),
    });

    // interpret response code. Log error for anything outside of 2xx 3xx
    // https://www.mongodb.com/docs/atlas/api/data-api-resources/#error-codes
    if (response.status < 200 || response.status >= 400) {
      let errorText: string | undefined;
      try {
        errorText = ((await response.json()) as { error: string })?.error;
      } catch {
        /* c8 ignore next */
      }

      const fallbackMessage = {
        400: "Bad Request",
        401: "Unauthorized",
        404: "Not Found",
        500: "Internal Server Error",
      };

      return {
        error: new MongoDataAPIError(
          errorText ??
            response.statusText ??
            fallbackMessage?.[response.status] ??
            "Data API Error",
          response.status
        ),
      };
    }

    const text = await response.text();

    return { data: EJSON.parse(text) as T };
  }
}
